python/metatensor-learn/metatensor/learn/nn/module_map.py

```diff
@@ -0,0 +1,470 @@
+        ...                 [0, 0],
+        ...                 [0, 1],
+        ...             ]
+        ...         ),
+        ...     ),
+        ...     components=[],
+        ...     properties=Labels(["properties"], np.array([[0], [1], [2]])),
+        ... )
+        >>> keys = Labels(names=["key"], values=np.array([[0], [1]]))
+        >>> tensor = TensorMap(keys, [block_1, block_2])
+        >>> linear = torch.nn.Linear(3, 1, bias=False)
+        >>> with torch.no_grad():
+        ...     _ = linear.weight.copy_(torch.tensor([1.0, 1.0, 1.0]))
+        ...
+        >>> # you could also extend the module by some nonlinear activation function
+        >>> from metatensor.learn.nn import ModuleMap
+        >>> module_map = ModuleMap.from_module(tensor.keys, linear)
+        >>> out = module_map(tensor)
+        >>> out[0].values
+        tensor([[ 7.],
+                [14.]], grad_fn=<MmBackward0>)
+        >>> out[1].values
+        tensor([[15.],
+                [11.]], grad_fn=<MmBackward0>)
+        """
+        module = deepcopy(module)
+        modules = []
+        for _ in range(len(in_keys)):
+            if many_to_one:
+                modules.append(module)
+            else:
+                modules.append(deepcopy(module))
+
+        return cls(in_keys, modules, out_tensor)
+
+    def forward(self, tensor: TensorMap) -> TensorMap:
+        """
+        Takes a tensor map and applies the modules on each key it.
+
+        :param tensor:
+            input tensor map
+        """
+        out_blocks: List[TensorBlock] = []
+        for key, block in tensor.items():
+            out_block = self.forward_block(key, block)
+
+            for parameter, gradient in block.gradients():
+                if len(gradient.gradients_list()) != 0:
+                    raise NotImplementedError(
+                        "gradients of gradients are not supported"
+                    )
+                out_block.add_gradient(
+                    parameter=parameter,
+                    gradient=self.forward_block(key, gradient),
+                )
+            out_blocks.append(out_block)
+
+        return TensorMap(tensor.keys, out_blocks)
+
+    def forward_block(self, key: LabelsEntry, block: TensorBlock) -> TensorBlock:
+        position = self._in_keys.position(key)
+        if position is None:
+            raise KeyError(f"key {key} not found in modules.")
+        module_idx: int = position
+        module: ModuleMapInterface = self._modules_list[module_idx]
+        out_values = module.forward(block.values)
+        if self._out_tensor is None:
+            properties = Labels.range("_", out_values.shape[-1])
+        else:
+            properties = self._out_tensor.block(key).properties
+        return TensorBlock(
+            values=out_values,
+            properties=properties,
+            components=block.components,
+            samples=block.samples,
+        )
+
+    @property
+    def modules(self):
```

Comment:
AFAICS, this property method is never used in a code. Should we still keep it here?

It seems, that there may be two purposes for such a method: 
1. If we want to create a shortcut for `self._modules_list`
2. If we want to display the modules used inside a `ModuleMap`

So the first feature is not used anyway. As for the second - we can just see the modules with the basic printouts (`__str__()` and `__repr__()`)

Reply:
So the only purpose of this method is to provide the user with a safe way of having an access to the modules list?